### ISSUE TITLE
250917 Generation

### ISSUE THREAD

**Issue by suhwankimkim** [AIBootcamp14/wiki-rider#47]
## 📅 학습 날짜  
* 2025-09-17

## 📌 목차
1. **GAN**
    *  What is GANs?
    * GAN VS VAE
    * GAN 구조
    * GAN 목적함수
    * GAN 학습
    * GAN 한계점
  
2. **조건부 생성 모델**
    * What is a 조건부 생성 모델?
    * 조건부 GAN
    * 다중 모드 기반 조건부 생성

3. **조건부 생성 모델을 활용한 다양한 영상 조작**
    * 이미지 vs 이미지
    * 텍스트 vs 이미지

----
  
### **1.1 What is GANs?**
>  GAN은 Generative Adversarial Networks에 줄임말로서 한국어로 적대적 생성 신경망을 말한다. 

>  생성자(Generator)와 판별자(Discriminator)로 구성되어져 있으며, 이미지가 주어졌을 때 판별자(Discriminator)는 진짜 이미지인지 가짜 이미지인지 구분하는 역할, 생성자는 판별자가 가짜와 진짜를 구분하지 못할 수준의 이미지를 만들어내는 역할으로서 서로를 견제하며 학습하는 구조에 신경망이다.

<br>

### **1.2 GAN vs VAE**
* 생성 방식 관점에서의 차이
  1. VAE : 입력 분포를 근사하는 방식으로 학습 및 데이터 생성
  2. GAN : 생성 이미지와 실제 이미지를 판별(Discriminator)하고 속이는(Generator) 과정을 거치며 생성자(Generator) 학습 및 데이터 생성

<br>

* 생성 결과물 관점에서의 차이

|  |VAE|GAN|
|:---:|----------------|-------------------------|
|결과물의 특징|GAN에 비해 상대적으로 결과물이 **흐릿**하다.|VAE에 비해 상대적으로 결과물이 **뚜렷**하다.|
|입력 데이터와의 관계|입력 데이터와 **유사한 형태**로 생성|입력 데이터와 **다른 형태의 데이터**를 생성|

<br>

### **1.3 GAN 구조**
> 데이터를 생성하는 생성자(Generator)와 입력받은 데이터가 가짜인지 진짜인지 판별하는 판별자(Discriminator)로 구성되어져 있다.

* 생성자(Generator) : 임의의 Noise를 입력 받아 가짜 이미지 생성
* 판별자(Discriminator) : 생성자가 만든 이미지 또는 실제 이미지를 입력 받아 진위 여부를 판단

> 끝없이 생성자와 판별자가 서로 속이고 잡아내고를 통해 싸우면서 성장, 이러한 학습과정이 다른 신경망 모형들과 다른 특이한 구조로 이루어져 있어 적대적(Adversarial)이라는 단어가 모형이름에 붙음

<br>

### **1.4 GAN 목적함수**

Objective function : $\min_{G}\max_{D} V(D,G) =E_{x \sim p_{data}(x)}[\log{D(x)}] + E_{z \sim p_{z}(z)}[\log(1-D(G(z)))].$

1. 고정된 G에서의 최적 $D^*(x) = \frac{p_{data}(x)}{p_{data}(x)+p_{G}(x)}.$
    > $\because\ f = a\log D+b\log(1-D),$
    > $\ f' = \frac{a}{D} - \frac{b}{1-D},$
    > $\ D(b-a) + a = 0, D= \frac{a}{a-b} \  \because f \ is \ concave \ function$
    > $\ \therefore, D^*(x) = \frac{p_{data}(x)}{p_{data}(x)+p_{G}(x)}.$

2. 최적 D를 대입한 값 $V(D^*,G) = -\log 4 + 2JSD(p_{data}\|p_G).$
    > JSD(Jensen-Shannon Divergence)는 KL과 맞찬가지로 서로 다른 확률분포의 다른 정도를 재는 지표. KL 발산을 대칭, 안정화한 버전의 지표.
    > $JSD(P\|Q)=\frac{1}{2}KL(P\|M)+\frac{1}{2}KL(Q\|M), \quad M=\frac{1}{2}(P+Q)$

3. 전역최적
    > $p_{G} = p_{data} \Rightarrow JSD = 0, \quad V( D^{ * }, V^{ * } )= -\log 4, \quad D^{*}(x) =  \frac{1}{2}, \ \forall x$

* 정리하면, 판별자(Discriminator)는 Objective function $V(D,G)$를 최대화 하도록 학습, 생성자(Generator)는 목적함수를 최소화 하도록 학습.
* 즉, 서로 다른 방향으로 학습(적대적 학습)하면서 판별자는 더욱 교묘한 fake, real을 구분을 잘하는 방향으로 생성자는 더욱 진짜와 같이 생성하도록 학습.
* 전역 최적은 $p_g = p_{data}$, 즉 생성자의 분포가 실제 데이터의 분포에 근사해야 이론적인 목적함수의 최적값 $-\log 4$로 수렴하게 된다.

<br>

### **1.5 GAN 학습**
**GAN 학습 방식**
* 생성자와 판별자 두 모델이 적대적인 방향으로 훈련 하게 된다.
    1. 판별 모델의 기울기 상승 : $max_D(E_{x \sim p_{data}}[\log D(x)]+E_{z \sim p_z}[log(1-D(G(z))])$
    2. 생성 모델의 기울기 하강 : $min_G E_{z \sim p_z}[\log(1-D(G(z)))]$
    > 정리하자면 Binary Cross Entropy와 매우 유사하게 생겼다. 판별자가 real image라고 판단한 확률의 기댓값과 fake image라고 판단한 확률의 기댓값의 합으로 표현된 것이라고 볼 수 있다.

    > 판별 모델의 학습은 주어진 목적함수의 최대화를 목표하기 때문에 $D(x)$는 최대한 1의 가깝게, $1-D(G(z))$도 최대한 1의 가깝게 즉 fake image에 대해서 0이라고 예측 할 수 있게 판별모델이 학습한다는 것을 의미한다.

    > 생성 모델의 학습은 목적함수의 최소화를 목표하기 때문에 $1-D(G(z))$가 최대한 0의 가깝게 즉 판별자가 real이라고 판별하게 G가 학습한다는 것을 의미한다.

* GAN 학습 방식 **문제점**
    1. 생성자 학습에 사용되는 목적함수의 구조가 학습이 잘 이루어졌을 때 기울기가 커지고, 학습이 잘 이루어지지 않았을 때 기울기가 작아지는 경향이 있다. 
    > 이를 해결하기 위해 $D(G(z))$의 값을 1에 가깝게 학습하는 방향은 유지하면서 기울기와 관련한 문제 해결을 위해 목적함수를 개선 할 필요가 있다.
    > $min_G E_{z \sim p_z}[ \log (1-D(G(z)))] \ \Rightarrow \ min_G E_{z \sim p_z}[-log(D(G(z)))]$ 로 개선하여 문제 해결

    2. 모드 붕괴(Mode Collapse) 현상이 발생할 수 있다.
    > 모드 붕괴(Mode Collapse)란? 생성자(Generator)가 다양성을 제대로 펼치지 않고, 소수(심하면 한 개)의 패턴/모드만 반복 생성하는 현상을 말한다. 즉 모델분포 $p_G$가 실제 분포 $p_{data}$의 여러 봉우리(모드) 중 일부만 커버하는 경우에 발생한다.

* 생성자와 판별자 두 개의 모델을 학습하는 것은 굉장히 **불안정**하며 여전히 풀어야만 하는 문제 학습에 유리한 목적 함수를 찾는 것이 중요
    > 다양한 Objective function을 일반화한 f-GANs라는 논문도 존재한다.

<details>
  <summary>
<span style="font-weight: bold; color:crimson; background-color: #fce4ec;">
모드 붕괴 주요 발생원인
</span>
</summary>

1. **목적함수/기울기 소실 문제**  
   GAN의 목적함수는 판별자가 최적이라고 가정하면 $-\log 4 + 2 \cdot JS(p_{data} \| p_G)$ 이다. 즉, GAN은 사실상 생성분포와 실제분포 사이의 JS-divergence를 최소화하려는 구조라고 이해할 수 있다. 그런데 JS-divergence는 두 분포가 멀리 떨어져 있을 때 **기울기 소실(gradient vanishing)** 문제가 발생한다.

2. **판별자 D와 생성자 G의 사이클/추격(chasing)**  
   판별자 D의 성능이 너무 강하거나(overfitting), 학습률/업데이트 횟수 불균형(G 1번 update ↔ D 여러번 update)일 경우  
   G가 가장 속이기 쉬운 좁은 영역으로만 수렴하게 됨 ➡ 모드 붕괴

3. **목적함수 다양성 부족**  
   GAN 목적함수는 "real처럼 보이게"만 최적화하고, 샘플 간 **다양성 보상은 없음** → 여러 z가 같은 출력으로 매핑되어도 목적은 달성됨.  
   (예: 고양이 데이터에서 앉은 고양이만 생성)

4. **Latent dimension / batch size / G 용량 부족**

5. **클래스 불균형 / 데이터 다양성 부족**  
   ➡ 생성자 G가 쉬운 모드에만 집착
</details>

### **1.6 GAN 한계점과 의의**
* 두 신경망이 적대적으로 학습되는 형식으로 인해 기존 VAE보다 상대적으로 선명한 이미지를 생성
* 최적화가 매우 어렵고 두 개의 신경망을 학습해야 하므로 학습이 오래 걸린다.
* 생성된 데이터의 다양성(모드 붕괴)이 낮을 수 있따.

<br>

---

### **2.1 What is a 조건부 생성 모델?**
* 조건부 생성모델 발생 배경
> 기존에 생성 모델들은 임의의 잠재 벡터로부터 데이터를 생성이 가능하지만 생성된 이미지에 의미는 제어 할 수 없다는 문제가 있다. 즉 다양한 활용(데이터 증강, 영상 편집, 간단한 그림 등)을 위해 제어를 할 수 없다는 문제들이 있었다. 이러한 문제점들을 해결 하기 위해 고안된 모델

* **What is a 조건부 생성 모델?**
> 임의의 잠재 벡터 + 조건 정보를 추가하여 데이터를 생성하는 방식, 예를 들어 MNIST 이미지와 유사한 0이라는 숫자 이미지를 만들고 싶을 때 숫자 0이라는 조건부 벡터를 임의의 잠재벡터에 추가해 model에 input 해서 0과 관련된 이미지만 추출

### **2.2 조건부 GAN**
**조건부 GAN**
* 생성자(Generator)에 입력되는 Latent vector와 판별자(Discriminator)에 입력되는 input image $x$에 대해 조건부 벡터가 추가되어서 학습하게 된다.
* Objective function : $\min_{G}\max_{D} V(D,G) =E_{x \sim p_{data}(x)}[\log{D(x|c)}] + E_{z \sim p_{z}(z)}[\log(1-D(G(z|c)))].$
* 판별자(Discriminator)이 입력받은 데이터가 실제데이터와 유사하여도 입력된 조건을 만족하지 않으면 0을 출력한다.

### **2.3 다중 모드 기반 조건부 생성**
* 판별자(Discriminator)가 입력되어지는 조건을 어떻게 고려하는지에 따른 다양한 방식이 있고 방식에 따라 ACGAN, ProjGAN, ContraGAN 등 다양한 종류가 존재한다.

* 생성자(Generator)도 역시 판별자와 마찬가지로 입력되어지는 조건을 어떻게 입력받는지에 따라 여러 방식이 있다.

* 조건은 나비, 인간, 나이와 같은 범주를 표현하는 one-hot encoding vector뿐 아니라, text, image, audio 등 다양한 형식의 조건을 입력받아 생성하는 방식으로 생성모델은 발전하고 있다.

**다양한 모드 기반의 조건부 생성모델**
1. Pix2pix(2017) : image to image
2. LostGANs(2019) : Layout to image
3. Speech2image(2020) : audio to image
4. GigaGAN(2023) : text to image

### **다조건부 생성 모델에 의의와 한계점**
* 조건을 입력하여 원하는 의미를 갖는 이미지를 생성할 수 있게 되었다.
* 범주, 오디오, text에 이르기까지 다양한 입력으로 이미지를 생성할 수 있게 되었다.
* 수집하기 매우 어려운 데이터를 학습에 요구됨

## ❓ 어려웠던 부분  
- 고도화된 생성 모델들이 이해하는데 어려움이 생김 pix2pix에 주제에 맞는 이미지 쌍(흑백 vs 컬러, semantic segmentation image vs original image, winter vs summer etc...) 이 이미지 쌍들이 어떻게 input으로 모델에 넣어져 학습하는지에 대한 이해가 어려움. 기존 판별자가 G(z)와 실제 이미지인 x를 잘 구분하는 task에서 어떤식으로 바뀌었는지 이해가 잘되지 않음.
- StarGAN은 정해두었던 도메인에 대한 변환이 이루어지는 것인가?
- 목적함수에 대한 직관적인 이해가 고도화된 GAN에서는 어려움이 있었음

**Comment by LeeHyunJong** [comment]
GANs 관련 이론을 자세히 잘 정리해주셨습니다! GANs는 결국 Jensen-Shannon (JS) 발산을 최소화하는 것으로 생각할 수 있습니다. 정리해주신대로 JS 발산은 기울기 소실 문제가 발생할 수 있어, 이를 개선한 모형이 [Wasserstein GANs](https://arxiv.org/pdf/1701.07875) 입니다. GANs 모형 중 유명한 모델 중 하나이므로 관심 있으시면 살펴보면 좋을 것 같습니다.


### 어려웠던 부분

> 고도화된 생성 모델들이 이해하는데 어려움이 생김 pix2pix에 주제에 맞는 이미지 쌍(흑백 vs 컬러, semantic segmentation image vs original image, winter vs summer etc...) 이 이미지 쌍들이 어떻게 input으로 모델에 넣어져 학습하는지에 대한 이해가 어려움. 기존 판별자가 G(z)와 실제 이미지인 x를 잘 구분하는 task에서 어떤식으로 바뀌었는지 이해가 잘되지 않음.

Pix2pix는 이미지 자체를 생성 모델의 입력으로 사용합니다. 이를 위해 U-net 관련 구조를 활용합니다. 그리고 랜덤한 생성을 위해 노이즈를 입력해주는 것이 필요한데, 이미지와 함께 랜덤 벡터를 입력으로 넣으면 랜덤 벡터를 무시해버리는 경향이 있어 생성모델에 Dropout 레이어를 활용하였습니다.

Pix2pix에서 활용하는 데이터는 이미지 쌍입니다. 판별 모델은 진짜 이미지 쌍(조건 이미지, 실제 결과)와 가짜 이미지 쌍(조건 이미지, 생성 모델의 출력)을 구분하도록 설계됩니다. 이미지 쌍은 channel 단위로 합쳐져 판별 모델의 입력으로 활용됩니다.

> StarGAN은 정해두었던 도메인에 대한 변환이 이루어지는 것인가?

이해하신 내용이 맞습니다. StarGAN은 주어진 도메인 레이블에 맞춰 이미지를 변환합니다. 도메인을 분류하는 분류기가 활용되기 때문에 주어진 도메인만 지원합니다. 이 점을 개선한 모델로는 [StarGAN v2](https://arxiv.org/pdf/1912.01865)가 있습니다.

> 목적함수에 대한 직관적인 이해가 고도화된 GAN에서는 어려움이 있었음

GAN이 발전할수록 목적함수가 복잡해지는데, 기존 GAN 손실에 모델 목적에 맞는 새로운 제약(regularization)을 추가한 것입니다. 이 구조에 맞춰서 해석하는 것이 도움이 됩니다. 예를 들면, 다음과 같습니다.
- Pix2pix: 기본 GAN 손실 + L1 손실(픽셀 단위 차이 최소화)
- CycleGAN: 기존 GAN 손실 + Cycle-consistency 손실 (+ identity 손실)
- StarGAN: 기본 GAN 손실 + 도메인 분류 손실 (변환된 이미지가 올바른 도메인에 속해야 함)
